experimental/slo/metrics.go

```diff
@@ -0,0 +1,227 @@
+package slo
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+// Metrics is an abstraction for collecting metrics
+type Metrics struct {
+	DSName   string
+	DSType   string
+	Endpoint Endpoint
+}
+
+// Duration is stored in the Context and used to collect metrics
+type Duration struct {
+	Value      float64
+	Status     Status
+	Source     Source
+	StatusCode int
+	mutex      sync.Mutex
+}
+
+func (d *Duration) Add(value float64, source Source, statusCode int, err error) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+	if d.Status == "" {
+		d.Status = "ok"
+	}
+	if err != nil {
+		d.Status = "error"
+	}
+	if statusCode >= 400 {
+		d.Status = "error"
+	}
+
+	// If the status code is now ok, but the previous status code was 401 or 403, mark it as ok
+	// assuming a successful re-authentication ( token refresh, etc )
+	if statusCode < 400 && (d.StatusCode == 401 || d.StatusCode == 403) {
+		d.Status = "ok"
+	}
+
+	d.StatusCode = statusCode
+	d.Source = source
+	d.Value += value
+}
+
+// Status is the status of the request
+type Status string
+
+// Endpoint is the endpoint of the request (health, query, resource)
+type Endpoint string
+
+// Source is the source of the error (downstream, plugin)
+type Source string
+
+const (
+	StatusOK         Status   = "ok"
+	StatusError      Status   = "error"
+	EndpointHealth   Endpoint = "health"
+	EndpointQuery    Endpoint = "query"
+	EndpointResource Endpoint = "resource"
+	SourceDownstream Source   = "downstream"
+	SourcePlugin     Source   = "plugin"
+)
+
+var durationMetric = promauto.NewHistogramVec(prometheus.HistogramOpts{
+	Namespace: "plugins",
+	Name:      "plugin_request_duration_seconds",
+	Help:      "Duration of plugin execution",
+}, []string{"datasource_name", "datasource_type", "source", "endpoint", "status", "status_code"})
```

Comment:
Nit. Wonder if `status_code` is really needed, leave that up to you. But in general, feels like logs could be more suitable for details following our discussion from last week.